SDIOBlockDevice.cpp

```diff
@@ -0,0 +1,378 @@
+
+    _sectors = 0;
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
+{
+    //debug_if(SD_DBG, "read Card...\r\n");
+    lock();
+    if (isPresent() == false)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+    if (!is_valid_read(addr, size))
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_PARAMETER;
+    }
+
+    if (!_is_initialized)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_INIT;
+    }
+
+    uint32_t *buffer = static_cast<uint32_t *>(b);
+
+    // ReadBlocks uses byte unit address
+    // SDHC and SDXC Cards different addressing is handled in ReadBlocks()
+    bd_addr_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+    // make sure card is ready
+    while (SD_GetCardState() != SD_TRANSFER_OK)
+    {
+        // wait until SD ready
+        wait_ms(1);
+    }
+
+    // receive the data : one block/ multiple blocks is handled in ReadBlocks()
+    int status = SD_ReadBlocks_DMA(buffer, addr, blockCnt);
+    debug_if(SD_DBG, "ReadBlocks dbgtest addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+
+    if (status == MSD_OK)
+    {
+        // wait until DMA finished
+        while (SD_DMA_ReadPending() != SD_TRANSFER_OK)
+        {
+            uint32_t tickstart = HAL_GetTick();
+            if ((HAL_GetTick() - tickstart) >= MBED_CONF_SD_TIMEOUT)
+            {
+                unlock();
+                return SD_BLOCK_DEVICE_ERROR_READBLOCKS;
+            }
+        }
+        // make sure card is ready
+        while (SD_GetCardState() != SD_TRANSFER_OK)
+        {
+            // wait until SD ready
+            wait_ms(10);
+        }
+    }
+    else
+    {
+        debug_if(SD_DBG, "ReadBlocks failed! addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+        debug_if(SD_DBG, "  hsd.errorcode: %lu  0x%lx\n", hsd.ErrorCode, hsd.ErrorCode);
+        status = SD_BLOCK_DEVICE_ERROR_READBLOCKS;
+    }
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::program(const void *b, bd_addr_t addr, bd_size_t size)
+{
+    //debug_if(SD_DBG, "program Card...\r\n");
+    lock();
+
+    if (isPresent() == false)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+    if (!is_valid_program(addr, size))
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_PARAMETER;
+    }
+
+    if (!_is_initialized)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_INIT;
+    }
+
+    //uint32_t *buffer = static_cast< uint32_t *>(b);
+    // HAL layer uses uint32_t for addr/size
+    uint32_t *buffer = (uint32_t *)(b);
+
+    // Get block count
+    bd_size_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+    // make sure card is ready
+    while (SD_GetCardState() != SD_TRANSFER_OK)
+    {
+        // wait until SD ready
+        wait_ms(1);
+    }
+
+    int status = SD_WriteBlocks_DMA(buffer, addr, blockCnt);
```

Comment:
same remark regarding DMA, unless all SDIO must work with DMA - it should be moved to an abstraction layer API

Reply:
same reason, DMA is neccessary